iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs

```diff
@@ -400,8 +401,13 @@ private async Task<Twin> RoundTripTwinMessageAsync(AmqpTwinMessageType amqpTwinM
                 await _amqpUnit.SendTwinMessageAsync(amqpTwinMessageType, correlationId, reportedProperties, _operationTimeout).ConfigureAwait(false);
 
                 var receivingTask = taskCompletionSource.Task;
+
                 if (await Task.WhenAny(receivingTask, Task.Delay(TimeSpan.FromSeconds(ResponseTimeoutInSeconds), cancellationToken)).ConfigureAwait(false) == receivingTask)
                 {
+                    if ((receivingTask.Exception != null) && (receivingTask.Exception.InnerException != null))
+                    {
+                        throw receivingTask.Exception.InnerException;
```

Comment:
I really like the overall implementation, my only question is - by rethrowing the exception here do you think we'll lose the stack details?

Reply:
synced offline, stack details is not lost. :)